File: backend/app/api/data/income.py
"""소득·성장 (급여 상세·인상 시뮬·부업·투자수익) endpoints."""
from __future__ import annotations

from fastapi import APIRouter, Body, Depends, Query

from app.core.auth import require_auth

from app.data.market import income

router = APIRouter()


@router.get("/income/overview")
def income_overview(user: str = Depends(require_auth)):
    """소득 종합 — 급여·부업·투자수익 + 월 총소득 + 조언."""
    return income.overview(user)


@router.get("/income/salary")
def income_salary_get(user: str = Depends(require_auth)):
    return income.get_salary(user)


@router.post("/income/salary")
def income_salary_set(earnings: list[dict] = Body(...), deductions: list[dict] = Body(default=[]),
                      memo: str = Body(default=""), user: str = Depends(require_auth)):
    """지급/공제 항목별 급여 저장(월 기준). [{label, amount}]"""
    return income.set_salary(user, earnings, deductions, memo)


@router.get("/income/raise-sim")
def income_raise_sim(raise_pct: float = Query(default=0), raise_amount: float = Query(default=0),
                     years: int = Query(default=5, ge=1, le=40),
                     invest_ratio: float = Query(default=0.5, ge=0.0, le=1.0),
                     annual_return: float = Query(default=6.0, ge=0.0, le=30.0),
                     user: str = Depends(require_auth)):
    """급여 인상 시나리오 + 인상분 적립투자 복리 미래가치."""
    return income.raise_sim(user, raise_pct, raise_amount, years, invest_ratio, annual_return)


@router.get("/income/side")
def income_side_list(month: str | None = Query(default=None), user: str = Depends(require_auth)):
    return income.list_side(user, month)


@router.post("/income/side")
def income_side_add(items: list[dict] = Body(...), user: str = Depends(require_auth)):
    """부업 소득 추가 [{date, source, amount, memo}]."""
    return income.add_side(user, items)


@router.post("/income/side/delete")
def income_side_delete(sid: int = Query(...), user: str = Depends(require_auth)):
    return income.delete_side(user, sid)

File: frontend/src/lib/api/income.ts
// 소득·성장 — 급여 상세/인상 이력·인상 시뮬·부업

import { request } from "./client";

export interface SalaryItem { label: string; amount: number }

export interface SalaryComputed {
  earnings: SalaryItem[];
  deductions: SalaryItem[];
  memo: string;
  gross: number;
  deduction: number;
  net: number;
  annual_net: number;
  annual_gross: number;
  updated: string | null;
}

export interface SalaryHistory { date: string; gross: number; net: number; annual_net: number }

export interface RaiseSim {
  base_net: number; new_net: number; monthly_increase: number; annual_increase: number;
  years: number; invest_ratio: number; annual_return: number;
  invest_monthly: number; contributed: number; future_value: number; investment_gain: number; note: string;
}

export interface SideRow { id: number; date: string; source: string; amount: number; memo: string }

export interface SideList {
  month: string | null; months: string[]; rows: SideRow[];
  month_total: number; total: number; sources: { source: string; amount: number }[];
}

export interface IncomeOverview {
  salary: SalaryComputed | null;
  side: { this_month: number; total: number; count: number };
  investment: { value: number; pnl: number; pnl_pct: number | null };
  total_month_income: number;
  annual_est: number;
  tips: string[];
}

export const incomeApi = {
  incomeOverview: () => request<IncomeOverview>(`/api/data/income/overview`),
  incomeSalaryGet: () => request<{ salary: SalaryComputed | null; history: SalaryHistory[] }>(`/api/data/income/salary`),
  incomeSalarySet: (earnings: SalaryItem[], deductions: SalaryItem[], memo = "") =>
    request<SalaryComputed>(`/api/data/income/salary`, { method: "POST", body: JSON.stringify({ earnings, deductions, memo }) }),
  incomeRaiseSim: (p: { raise_pct?: number; raise_amount?: number; years?: number; invest_ratio?: number; annual_return?: number }) => {
    const q = new URLSearchParams();
    if (p.raise_pct != null) q.set("raise_pct", String(p.raise_pct));
    if (p.raise_amount != null) q.set("raise_amount", String(p.raise_amount));
    if (p.years != null) q.set("years", String(p.years));
    if (p.invest_ratio != null) q.set("invest_ratio", String(p.invest_ratio));
    if (p.annual_return != null) q.set("annual_return", String(p.annual_return));
    return request<RaiseSim>(`/api/data/income/raise-sim?${q.toString()}`);
  },
  incomeSideList: (month?: string) => request<SideList>(`/api/data/income/side${month ? `?month=${month}` : ""}`),
  incomeSideAdd: (items: { date: string; source: string; amount: number; memo?: string }[]) =>
    request<{ added: number }>(`/api/data/income/side`, { method: "POST", body: JSON.stringify(items) }),
  incomeSideDelete: (sid: number) => request<{ ok: boolean }>(`/api/data/income/side/delete?sid=${sid}`, { method: "POST" }),
};
